pytorch_lightning/metrics/functional/classification.py

```diff
@@ -632,6 +638,87 @@ def _multiclass_auroc(pred, target, sample_weight, num_classes):
     return torch.mean(class_aurocs)
 
 
+def object_detection_mean_average_precision(
+        preds: torch.Tensor, target: torch.Tensor, iou_threshold: float, ap_calculation: str
+) -> torch.Tensor:
+    """
+    Compute mean average precision for object detection task. Calculates average precision
+    using AUC.
+
+    Args:
+        preds: an Nx7 batch of predictions with representation
+               ``[image_idx, class_pred, class_prob, x_min, y_min, x_max, y_max]``
+        target: an Nx6 batch of targets with representation
+                ``[image_idx, class_label, x_min, y_min, x_max, y_max]``
+        iou_threshold: threshold for IoU score for determining true positive and
+                       false positive predictions.
+        ap_calculation: one of "step", "VOC2007", "VOC2012", or "COCO"
+
+    Returns:
+        mean of the average precision for each class in object detection task.
+    """
+    if box_iou is None:
+        raise ImportError('You want to use `torchvision` which is not installed yet,'
+                          ' install it with `pip install torchvision`.')
+    classes = torch.cat([preds[:, 1], target[:, 1]]).unique()
+    average_precisions = torch.zeros(len(classes))
+    for class_idx, c in enumerate(classes):
+        c_preds = sorted(preds[preds[:, 1] == c], key=lambda x: x[2], reverse=True)
+        c_target = target[target[:, 1] == c]
+        targets_per_images = Counter([t[0].item() for t in c_target])
+        targets_assigned = {
+            image_idx: torch.zeros(count, dtype=torch.bool) for image_idx, count in targets_per_images.items()
+        }
+        tps = torch.zeros(len(c_preds))
+        fps = torch.zeros(len(c_preds))
+        if len(c_preds) == 0:
+            continue
+        for i, p in enumerate(c_preds):
+            image_idx = p[0].item()
+            ground_truths = c_target[c_target[:, 0] == image_idx]
+            best_iou = 0
+            best_target_idx = 0
+            for j, t in enumerate(ground_truths):
+                iou = box_iou(p[None, 3:], t[None, 2:])
+                if iou > best_iou:
+                    best_iou = iou
+                    best_target_idx = j
+            if best_iou > iou_threshold and targets_assigned[image_idx][best_target_idx] == 0:
```

Comment:
Now you can write this (if I'm not mistaken):

            if best_iou > iou_threshold and not targets_assigned[image_idx][best_target_idx]:
               targets_assigned[image_idx][best_target_idx] = True